gxutil/pm.go

```diff
@@ -775,3 +670,208 @@ func setInstallPathCache(env string, global bool, val string) {
+	addedDepCount := 0
+	for _, dep := range pkg.Dependencies {
+		if dq.added[dep.Hash] == false {
+			dq.queue = append(dq.queue, dep)
+			addedDepCount++
+			dq.added[dep.Hash] = true
+		}
+	}
+	return addedDepCount
+}
+
+// Len returns the number of dependencies currently stored in the queue.
+func (dq *DependencyQueue) Len() int {
+	return len(dq.queue)
+}
+
+// GetDependency pops the first dependency in the queue and
+// returns it (or `nil` if the queue is empty).
+func (dq *DependencyQueue) GetDependency() *Dependency {
+	if dq.Len() == 0 {
+		return nil
+	}
+
+	dep := dq.queue[0]
+	dq.queue = dq.queue[1:]
+	return dep
+}
+
+// Fetch all of the dependencies of this package (direct and transitive
+// ones). Use (if possible) `maxGoroutines` goroutines working in parallel
+// (coordinated by this function). Each new dependency fetched is another
+// package with more (potentially new) dependencies that may also be fetched.
+//
+// TODO: Depending on the perspective sometimes we use the *package*
+// term and others *dependency* (of another package), that should
+// be unified and clarified as much as possible (not just in this function).
+func (pm *PM) fetchDependencies(pkg *Package, location string) error {
+
+	// Maximum number of goroutines allowed to run in parallel fetching
+	// packages.
+	const maxGoroutines = 20
+	// TODO: Consider making this value a parameter of the function
+	// (or an attribute of the `PM` structure).
+
+	// Central queue of dependencies that need to be fetched. Handled only
+	// by this function. Created with an initial a capacity on the rough
+	// estimate of twice the maximum goroutines running.
+	depQueue := NewDependencyQueue(maxGoroutines * 2)
+
+	// List of channels for each spawned goroutine to store either the
+	// fetched package or an error. To ensure they are non blocking the
+	// maximum number of goroutines it's assigned for their capacity
+	// (worst case scenario).
+	fetchedPackages := make(chan *Package, maxGoroutines)
+	fetchErrs := make(chan error, maxGoroutines)
+
+	// To start the process add the dependencies of the root package.
+	addedDepCount := depQueue.AddPackageDependencies(pkg)
+	pm.ProgMeter.AddTodos(addedDepCount)
+
+	// Counter to keep track of spawned goroutines, it's not locked as it's
+	// only handled by this function. Decremented any time a message is received
+	// on the above channels, which indicates that the goroutine has finished.
+	activeGoroutines := 0
+
+	// Main loop of the function.
+	for {
+		// If there are no more dependencies to install and there aren't any
+		// goroutines running (which could potentially add new dependencies
+		// to the queue) we're finished.
+		if depQueue.Len() == 0 && activeGoroutines == 0 {
+			return nil
+		}
+
+		// Keep spawning goroutines until the allowed maximum or until
+		// there are no new dependencies to fetch at the moment.
+		for activeGoroutines < maxGoroutines && depQueue.Len() > 0 {
```

Comment:
Nit: I *usually* do this by spawning N goroutines up-front, having them select on a context cancellation and a work channel. You can then use a `sync.WaitGroup` to, e.g., wait for all the active goroutines to exit when the context is canceled. This also saves the GC some work.

However, the code is clean as-is so there's no *need* to do it this way.

Reply:
Could you point me to an example of that pattern in the code base to have a reference? My main motivation here is to have the code as simple as possible (the bottleneck here is, by far, the network I/O), so if it seems simpler than this I'll use that.

Reply:
Thanks for the reference I'll keep it in mind.

The problem here is that the work (each fetch) is not independent of each other, I need the result of that work to decide if I give more work to the workers (new dependencies -of the already fetched ones- to be fetched) so I need some control mechanism to decide when to finish (because I need to ask the worker: are you actually fetching something right now? in that case I'll wait for that output, or if you're idle -all of them are idle- then I can close the channel and finish).

Reply:
The usual solution to that is usually counting (count outstanding tasks). But really, the current solution is just fine.

Reply:
I see, yes, I was counting goroutines when I should have been counting job requests.